libkbfs/root_metadata.go

```diff
@@ -47,288 +45,22 @@ func (p PrivateMetadata) checkValid() error {
-	MetadataFlagWriterMetadataCopied
-	MetadataFlagFinal
-)
-
-// WriterFlags bitfield.
-type WriterFlags byte
-
-// Possible flags set in the WriterFlags bitfield.
-const (
-	MetadataFlagUnmerged WriterFlags = 1 << iota
-)
-
-// MetadataRevision is the type for the revision number.
-// This is currently int64 since that's the type of Avro's long.
-type MetadataRevision int64
-
-// String converts a MetadataRevision to its string form.
-func (mr MetadataRevision) String() string {
-	return strconv.FormatInt(mr.Number(), 10)
-}
-
-// Number casts a MetadataRevision to it's primitive type.
-func (mr MetadataRevision) Number() int64 {
-	return int64(mr)
-}
-
-const (
-	// MetadataRevisionUninitialized indicates that a top-level folder has
-	// not yet been initialized.
-	MetadataRevisionUninitialized = MetadataRevision(0)
-	// MetadataRevisionInitial is always the first revision for an
-	// initialized top-level folder.
-	MetadataRevisionInitial = MetadataRevision(1)
-)
-
-// WriterMetadata stores the metadata for a TLF that is
-// only editable by users with writer permissions.
-//
-// NOTE: Don't add new fields to this type! Instead, add them to
-// WriterMetadataExtra. This is because we want old clients to
-// preserve unknown fields, and we're unable to do that for
-// WriterMetadata directly because it's embedded in RootMetadata.
-type WriterMetadata struct {
-	// Serialized, possibly encrypted, version of the PrivateMetadata
-	SerializedPrivateMetadata []byte `codec:"data"`
-	// The last KB user with writer permissions to this TLF
-	// who modified this WriterMetadata
-	LastModifyingWriter keybase1.UID
-	// For public TLFs (since those don't have any keys at all).
-	Writers []keybase1.UID `codec:",omitempty"`
-	// For private TLFs. Writer key generations for this metadata. The
-	// most recent one is last in the array. Must be same length as
-	// RootMetadata.RKeys.
-	WKeys TLFWriterKeyGenerations `codec:",omitempty"`
-	// The directory ID, signed over to make verification easier
-	ID TlfID
-	// The branch ID, currently only set if this is in unmerged per-device history.
-	BID BranchID
-	// Flags
-	WFlags WriterFlags
-	// Estimated disk usage at this revision
-	DiskUsage uint64
-
-	// The total number of bytes in new blocks
-	RefBytes uint64
-	// The total number of bytes in unreferenced blocks
-	UnrefBytes uint64
-
-	Extra WriterMetadataExtra `codec:"x,omitempty,omitemptycheckstruct"`
-}
-
-// WriterMetadataExtra stores more fields for WriterMetadata. (See
-// WriterMetadata comments as to why this type is needed.)
-type WriterMetadataExtra struct {
-	UnresolvedWriters []keybase1.SocialAssertion `codec:"uw,omitempty"`
-	codec.UnknownFieldSetHandler
-}
-
-// RootMetadata is the MD that is signed by the reader or writer.
+// A RootMetadata is a BareRootMetadata but with a deserialized
+// PrivateMetadata. However, note that it is possible that the
+// PrivateMetadata has to be left serialized due to not having the
+// right keys.
```

Comment:
On 07/15/2016 10:09 AM, Fred Akalin wrote:
>
> Yeah, that's possible. It'll probably be a yuuuge refactor though! :)
>


It would be the best refactor.  The greatest refactor the world has ever
seen.  No one else on earth could make a refactor as great as this one.